pkg/cmd/roachtest/tests/cdc.go

```diff
@@ -66,247 +67,403 @@ const (
 		if err != nil {
-			t.Fatal(err)
+			t.rt.Fatal(err)
 		}
 		sinkDest, err := cdctest.StartMockWebhookSink(cert)
 		if err != nil {
-			t.Fatal(err)
+			t.rt.Fatal(err)
 		}
 
 		sinkDestHost, err := url.Parse(sinkDest.URL())
 		if err != nil {
-			t.Fatal(err)
+			t.rt.Fatal(err)
 		}
 
 		params := sinkDestHost.Query()
 		params.Set(changefeedbase.SinkParamCACert, certEncoded)
 		sinkDestHost.RawQuery = params.Encode()
 
 		sinkURI = fmt.Sprintf("webhook-%s", sinkDestHost.String())
-	} else if args.whichSink == pubsubSink {
+	} else if whichSink == pubsubSink {
 		sinkURI = changefeedccl.GcpScheme + `://cockroach-ephemeral` + "?AUTH=implicit&topic_name=pubsubSink-roachtest&region=us-east1"
-	} else {
-		t.Status("installing kafka")
-		kafka.install(ctx)
-		kafka.start(ctx)
-		sinkURI = kafka.sinkURL(ctx)
+	} else { // Default to kafka
+		kafkaNode := t.cluster.Node(t.cluster.Spec().NodeCount)
+		kafka := kafkaManager{
+			rt:    t.rt,
+			c:     t.cluster,
+			nodes: kafkaNode,
+		}
+		kafka.install(t.ctx)
+		kafka.start(t.ctx)
+
+		if args.kafkaChaos {
+			t.m.Go(func(ctx context.Context) error {
+				period, downTime := 2*time.Minute, 20*time.Second
+				return kafka.chaosLoop(ctx, period, downTime, t.doneCh)
+			})
+		}
+
+		sinkURI = kafka.sinkURL(t.ctx)
 	}
 
 	if args.assumeRole != "" {
 		sinkURI = sinkURI + "&ASSUME_ROLE=" + args.assumeRole
 	}
 
-	m := c.NewMonitor(ctx, crdbNodes)
+	return sinkURI
+}
+
+func (t *cdcTester) setupAndRunWorkload(args workloadArgs) {
+	t.cluster.Put(t.ctx, t.rt.DeprecatedWorkload(), "./workload", t.workloadNode)
+
 	workloadCompleteCh := make(chan struct{}, 1)
+	t.workloadCompleteCh = workloadCompleteCh
 
-	workloadStart := timeutil.Now()
 	if args.workloadType == tpccWorkloadType {
-		t.Status("installing TPCC")
+		t.rt.Status("installing TPCC workload")
 		tpcc := tpccWorkload{
-			sqlNodes:           crdbNodes,
-			workloadNodes:      workloadNode,
+			sqlNodes:           t.crdbNodes,
+			workloadNodes:      t.workloadNode,
 			tpccWarehouseCount: args.tpccWarehouseCount,
 			// TolerateErrors if crdbChaos is true; otherwise, the workload will fail
 			// if it attempts to use the node which was brought down by chaos.
-			tolerateErrors: args.crdbChaos,
+			tolerateErrors: t.tolerateNodeErrors,
 		}
 
-		tpcc.install(ctx, c)
+		if !t.rt.SkipInit() {
+			tpcc.install(t.ctx, t.cluster)
+		}
 		// TODO(dan,ajwerner): sleeping momentarily before running the workload
 		// mitigates errors like "error in newOrder: missing stock row" from tpcc.
 		time.Sleep(2 * time.Second)
-		t.Status("initiating workload")
-		m.Go(func(ctx context.Context) error {
+		t.rt.Status("initiating TPCC workload")
+		t.m.Go(func(ctx context.Context) error {
 			defer func() { close(workloadCompleteCh) }()
-			tpcc.run(ctx, c, args.workloadDuration)
+			tpcc.run(ctx, t.cluster, args.duration)
 			return nil
 		})
-	} else {
-		t.Status("installing Ledger Workload")
+	} else if args.workloadType == ledgerWorkloadType {
```

Comment:
wdyt about moving these if/else branches to their own helpers (e.g. runLedgerWorkload, runTPCCWorkload)?